### PR TITLE
Type parameters as constraints

### LINKED ISSUE #2304: Allow type parameter constraints to reference other type parameters
I would like to have some feature that infers any possible types from the given object in the generic function.
For instance I have a function that receives Object, and I want to have a more specific callback than 'any'.
Current code:
```typescript
function a(obj: Object, f: (value: any) => any) {
 for(var p in obj) if(obj.hasOwnProperty(p)) {
  (<any>obj)[p] = f((<any>obj)[p]) 
 }
}

a({x:1}, (value: any) => (<number>value)+1);
```
I wish I had something with more strict typing.

```typescript
function a<T extends Object, Value: union_of_properties_types_of(T)>(obj: T, f: (value: Value) => Value) {
 for(var p in obj) if(obj.hasOwnProperty(p)) {
  (<any>obj)[p] = f(<Value>(<any>obj)[p]) 
 }
}

// 'f' is denoted to number => number
a({x:1}, value => value+1);

// 'f' is string => string
a({x:"x", y:"y"}, value => {
 return value+"z";
});
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR it becomes possible for a type parameter constraint to reference type parameters from the same type parameter list (this was previously an error). For example:

```typescript
function assign<T extends U, U>(target: T, source: U): T {
    for (let id in source) {
        target[id] = source[id];
    }
    return target;
}

let x = { a: 1, b: 2, c: 3, d: 4 };
assign(x, { b: 10, d: 20 });
assign(x, { e: 0 });  // Error
```

A fancy term for this capability is [F-Bounded Polymorphism](https://en.wikipedia.org/wiki/Bounded_quantification#F-bounded_quantification).

Fixes #2304.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4751,19 +4718,21 @@ namespace ts {
...
         }
 
-        function createInferenceMapper(context: InferenceContext): TypeMapper {
-            const mapper: TypeMapper = t => {
-                for (let i = 0; i < context.typeParameters.length; i++) {
-                    if (t === context.typeParameters[i]) {
-                        context.inferences[i].isFixed = true;
-                        return getInferredType(context, i);
+        function getInferenceMapper(context: InferenceContext): TypeMapper {
+            if (!context.mapper) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Invert the condition to reduce nesting.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -4751,19 +4718,21 @@ namespace ts {
...
+                const mapper: TypeMapper = t => {
+                    for (let i = 0; i < context.typeParameters.length; i++) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Consider extracting `context.typeParameters` out to a variable and caching the length

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -4148,7 +4165,13 @@ namespace ts {
                     type.constraint = targetConstraint ? instantiateType(targetConstraint, type.mapper) : noConstraintType;
                 }
                 else {
-                    type.constraint = getTypeFromTypeNode((<TypeParameterDeclaration>getDeclarationOfKind(type.symbol, SyntaxKind.TypeParameter)).constraint);
+                    const constraintDeclaration = getConstraintDeclaration(type);
+                    let constraint = getTypeFromTypeNode(constraintDeclaration);
+                    if (hasConstraintReferenceTo(constraint, type)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Can you rename `type` to `typeParam`?

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Not sure what you mean. There's a shared return statement at the bottom of the function and no else clause that would be reversed.

**Comment by ahejlsberg** [comment]
@mhegazy Want to comment on this one before I merge it?

File: src/compiler/checker.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I meant to just repeat the `return` statement, but it's not really a big deal.

**Comment by mhegazy** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
@ahejlsberg I tried this out and upon grabbing quick info on `a`, I'm seeing an uninstantiated type:

image


```ts
interface Mapper<T> {
    map<U extends T, V extends U[]>(f: (item: T) => U): V;
}

var m: Mapper<string>;
var a = m.map((x: string) => x);
```

I expected `a` to have the type `string[]`, so something isn't working out here.

**Comment by ahejlsberg** [comment]
@DanielRosenwasser Good catch. The problem is that the constraints aren't getting properly instantiated. Will put up a fix shortly.

**Comment by DanielRosenwasser** [comment]
Hey @ahejlsberg, @aleksey-bykov wrote a useful example on #6037 which demonstrates the use of type parameters being able to extend each other. Can we add the following as a test case?

```ts
function fold<a, r>(values: a[], result: r, fold: (result: r, value: a) => r): r {
    for (let value of values) {
        result = fold(result, value);
    }
    return result;
}

function append<a, b extends a>(values: a[], value: b): a[] {
    values.push(value);
    return values;
}


fold(
    [1, 2, 3],
    [] as [string, string][],
    (result, value) => append(
        result,
        ['', ''] as [string, string]
    )
);
```

**Comment by aleksey-bykov** [comment]
by the way
```typescript
// test.ts
export function append<a, b extends a>(result: a[], value: a): a[] {
    result.push(value);
    return result;
}
```

compiles with
```
node ./built/local/tsc.js --module commonjs test.ts 
```

but fails with
```
node ./built/local/tsc.js --module commonjs --declaration test.ts 
```
saying
```
test.ts(1,37): error TS4016: Type parameter 'b' of exported function has or is using private name 'a'.
```

**Comment by aleksey-bykov** [comment]
create a separate bug: #6040

**Comment by DanielRosenwasser** [comment]
I noticed the same thing on my bug where the type parameter was not being instantiated. Can you enable `// @declaration: true` for all these test cases?

**Comment by ahejlsberg** [comment]
Latest commits fix the uninstantiated type issue as well as issue in #6040.

**Comment by basarat** [comment]
Awesome :rose:. I've done aweful things in the past : https://github.com/TypeScriptBuilder/tsb/blob/24c5bbaa3e29587ab0830194c374cd7fa8bb2462/src/app/state/simpleRedux.ts#L90-L103 :)

**Comment by DanielRosenwasser** [comment]
@ahejlsberg I think it wouldn't be unreasonable to just get declaration emit for all of these.

**Comment by aleksey-bykov** [comment]
a poorman's implementation of this feature could have been done by passing
a function that maps A to B thus enforcing the constraint, consider:

foo <a, b extends a>(foo: a, bar:b): void;

vs

foo <a, b>(foo: a, bar: b, bIsA: (val: b) => a): void;

**Comment by isiahmeadows** [comment]
Question: does this patch support the following? [I've come across this pattern way too frequently](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern), and [I've needed it for way too many things](https://en.wikipedia.org/wiki/Bounded_quantification#F-bounded_quantification). In languages that support it, it's a very frequent idiom.

```ts
interface Type<T extends Type<T>> {}
```

**Comment by ahejlsberg** [comment]
@isiahmeadows Yes, that pattern (the "curiously recurring template pattern") is now supported. But also note that the [polymorphic 'this' type](https://github.com/Microsoft/TypeScript/pull/4910) is a more succinct way to accomplish the same thing.

**Comment by isiahmeadows** [comment]
I also came across an unusual use case where I needed two F-bounded type
parameters, which the `this` type doesn't cover. Given this supports that
as well, I'm happy.

**Pull request merged**